Objects/cown.c

```diff
@@ -0,0 +1,314 @@
+            PyErr_Format(PyExc_RegionError, "%s: %S", msg, o); \
+            return NULL; \
+        } \
+    } while(0);
+
+#define BAIL_UNLESS_ACQUIRED(o, msg) \
+    BAIL_UNLESS_OWNED(o, msg) \
+    BAIL_UNLESS_IN_STATE(o, Cown_ACQUIRED, msg)
+
+static PyObject *PyCown_acquire(PyCownObject *self) {
+    int expected = Cown_RELEASED;
+
+    // TODO: eventually replace this with something from pycore_atomic (nothing there now)
+    while (!atomic_compare_exchange_strong(&self->state._value, &expected, Cown_ACQUIRED)) {
+        sem_wait(&self->semaphore);
+        expected = Cown_RELEASED;
+    }
+
+    // Note: we must hold the GIL at this point -- note for future
+    // threading implementation.
+    PyThreadState *tstate = PyThreadState_Get();
+    self->owning_thread = tstate->thread_id;
+
+    Py_RETURN_NONE;
+}
+
+static PyObject *PyCown_release(PyCownObject *self) {
+    if (STATE(self) == Cown_RELEASED) {
+        BAIL_IF_OWNED(self, "BUG: Released cown had owning thread: %p");
+        Py_RETURN_NONE;
+    }
+
+    BAIL_UNLESS_OWNED(self, "Thread attempted to release a cown it did not own");
+
+    self->owning_thread = 0;
+    _Py_atomic_store(&self->state, Cown_RELEASED);
+    sem_post(&self->semaphore);
+
+    Py_RETURN_NONE;
+}
+
+int _PyCown_release(PyObject *self) {
+    PyObject* res = PyCown_release((PyCownObject *)self);
+    return res == Py_None ? 0 : -1;
+}
+
+int _PyCown_is_released(PyObject *self) {
+    PyCownObject *cown = (PyCownObject *)self;
+    return STATE(cown) == Cown_RELEASED;
+}
+
+static PyObject *PyCown_get(PyCownObject *self) {
+    BAIL_UNLESS_ACQUIRED(self, "Attempt to get value of unacquired cown");
+
+    if (self->value) {
+        return Py_NewRef(self->value);
+    } else {
+        Py_RETURN_NONE;
+    }
+}
+
+// Needed to test for region object
+extern PyTypeObject PyRegion_Type;
+extern PyTypeObject PyCown_Type;
+
+static PyObject *PyCown_set_unchecked(PyCownObject *self, PyObject *arg) {
+    // Cowns are cells that hold a reference to a bridge object,
+    // (or another cown or immutable object)
+    const bool is_region_object =
+        Py_IS_TYPE(arg, &PyRegion_Type) && _Py_is_bridge_object(arg);
+    if (is_region_object ||
+        arg->ob_type == &PyCown_Type ||
+        _Py_IsImmutable(arg)) {
+
+        PyObject* old = self->value;
+        Py_XINCREF(arg);
+        self->value = arg;
+
+        // Tell the region that it is owned by a cown,
+        // to enable it to release the cown on close
+        if (is_region_object) {
+            _PyRegion_set_cown_parent(arg, _PyObject_CAST(self));
+            if (_PyRegion_is_closed(arg)) {
+                PyCown_release(self);
+            } else {
+                _Py_atomic_store(&self->state, Cown_PENDING_RELEASE);
+                // TODO: do we need an owning thread for this?
+                PyThreadState *tstate = PyThreadState_Get();
+                self->owning_thread = tstate->thread_id;
```

Comment:
> // TODO: do we need an owning thread for this?

I would say so. The cown is writable in this state. It could be shared with other threads in this state, so we should make sure that only one can access the content.

Reply:
Thanks! I came to the same conclusion but forgot to remove the comment!